sdk/servicebus/service-bus/src/connectionContext.ts

```diff
@@ -64,6 +119,48 @@ export namespace ConnectionContext {
     const connectionContext = ConnectionContextBase.create(parameters) as ConnectionContext;
     connectionContext.clientContexts = {};
 
+    let waitForDisconnectResolve: () => void;
+    let waitForDisconnectPromise: Promise<void> | undefined;
+    Object.assign<ConnectionContext, ConnectionContextMethods>(connectionContext, {
+      isConnectionClosing() {
+        // When the connection is not open, but the remote end is open,
+        // then the rhea connection is in the process of terminating.
+        return Boolean(!this.connection.isOpen() && this.connection.isRemoteOpen());
+      },
+      async readyToOpenLink() {
+        log.error(`[${this.connectionId}] Waiting until the connection is ready to open link.`);
+        // Check that the connection isn't in the process of closing.
+        // This can happen when the idle timeout has been reached but
+        // the underlying socket is waiting to be destroyed.
+        if (this.isConnectionClosing()) {
+          // Wait for the disconnected event that indicates the underlying socket has closed.
+          await this.waitForDisconnectedEvent();
+        }
+        // Check if the connection is currently in the process of disconnecting.
+        if (waitForDisconnectPromise) {
+          // Wait for the connection to be reset.
+          await this.waitForConnectionReset();
+        }
+        log.error(`[${this.connectionId}] Connection is ready to open link.`);
+      },
+      waitForDisconnectedEvent() {
+        return new Promise((resolve) => {
+          log.error(
+            `[${this.connectionId}] Attempting to reinitialize connection` +
+              ` but the connection is in the process of closing.` +
+              ` Waiting for the disconnect event before continuing.`
+          );
+          this.connection.once(ConnectionEvents.disconnected, resolve);
+        });
+      },
+      waitForConnectionReset() {
+        if (waitForDisconnectPromise) {
```

Comment:
renaming this to 'waitFConnectionRefreshPromise` might help associate it with the function (since that's technically what we're waiting for?)

I think I misled myself a little because the first one is `waitForDisconnectedEvent` but really only cares that it fires.

Reply:
Renamed for clarity!